src/Hangfire.Mongo/Signal/Mongo/MongoSignal.cs

```diff
@@ -0,0 +1,159 @@
+using System.Threading;
+using System.Threading.Tasks;
+using Hangfire.Logging;
+using Hangfire.Mongo.Dto;
+using MongoDB.Driver;
+
+namespace Hangfire.Mongo.Signal.Mongo
+{
+
+    /// <summary>
+    /// Provide interprocess/server signalling
+    /// </summary>
+    internal class MongoSignal : IPersistentSignal
+    {
+
+        private static readonly ILog Logger = LogProvider.For<MongoSignal>();
+
+        private static readonly ConcurrentDictionary<string, EventWaitHandle> EventWaitHandles = new ConcurrentDictionary<string, EventWaitHandle>();
+
+        private IMongoCollection<SignalDto> _signal;
+
+        internal MongoSignal(IMongoCollection<SignalDto> signal)
+        {
+            _signal = signal ?? throw new ArgumentNullException(nameof(signal));
+        }
+
+        public void Listen(CancellationToken cancellationToken)
+        {
+            var options = new FindOptions<SignalDto>
+            {
+                // Our cursor is a tailable cursor and informs the server to await
+                CursorType = CursorType.TailableAwait,
+                // Make sure to time out once in a while to avoid complete blocking operation
+                MaxAwaitTime = TimeSpan.FromMinutes(1),
+            };
+
+            var filterBuilder = Builders<SignalDto>.Filter;
+            var filter = filterBuilder.Eq(s => s.Signaled, true);
+
+            using (var cursor = _signal.FindSync(filter, options, cancellationToken))
+            {
+                Logger.Debug($@"*** LISTEN ({Thread.CurrentThread.ManagedThreadId})");
+                cancellationToken.ThrowIfCancellationRequested();
+                cursor.ForEachAsync(document =>
+                {
+                    var signalDto = _signal.FindOneAndUpdate(
+                        s => s.Id == document.Id,
+                        Builders<SignalDto>.Update.Set(s => s.Signaled, false),
+                        null,
+                        cancellationToken);
+                    if (signalDto == null)
+                    {
+                        Logger.Debug($@"*** TRIGGERED ({Thread.CurrentThread.ManagedThreadId})");
+                    }
+                    else
+                    {
+                        Logger.Debug($@"*** OWNED: {signalDto.Name} ({Thread.CurrentThread.ManagedThreadId})");
+                    }
+                    EventWaitHandles
+                        .GetOrAdd(document.Name, n => new EventWaitHandle(false, EventResetMode.AutoReset))
+                        .Set();
+                }, cancellationToken).ContinueWith(t =>
+                {
+                    // TODO: Log if canceled?
+                    if (t.IsFaulted)
+                    {
+                        var messages = string.Join(Environment.NewLine, t.Exception.InnerExceptions.Select(e => e.Message));
+                        Logger.Debug($@"*** FAULT: {messages} ({Thread.CurrentThread.ManagedThreadId})");
+                    }
+                }).Wait();
```

Comment:
cursor is never exhausted and you dont need to re query?

Reply:
Are you sure? You need to show me...